### PR TITLE
[18.0][MIG] queue_job: Migrate to 18.0

### PR DESCRIPTION
**Pull request by thienvh332**
### Ported PRs

- #413
- #387
- #466
- #495
- #594
- #598
- #629
- #624
- #657
- #676
- #683

### Changes
- Cleaned TODO tags

- Refactor the JobSerialized field on top of odoo.fields.Json

    - Result
   
```
>>> self.env.cr.execute("SELECT column_name, data_type FROM information_schema.columns WHERE table_name = 'queue_job' AND column_name = 'kwargs';")
>>> self.env.cr.fetchall()
[('kwargs', 'jsonb')]

```
- Example for searching:
```
>>> self.env.cr.execute("SELECT kwargs->>'a' FROM queue_job WHERE kwargs ? 'a'")
>>> self.env.cr.dictfetchall()
[{'?column?': 'abc'}, {'?column?': 'abc'}]
```

### CODE REVIEW

**Comment by simahawk** [comment]
@thienvh332 can you please check pre-commit?
Also, I'd say that the test module should stay together w/ the main one https://github.com/OCA/queue/pull/694

File: queue_job/fields.py [hunk 1]
```diff
@@ -23,7 +24,7 @@ class JobSerialized(fields.Field):
     """
 
     type = "job_serialized"
-    column_type = ("text", "text")
+    _column_type = ("jsonb", "jsonb")
```
Comment (simahawk) [inline queue_job/fields.py:27]: to preserve existing jobs this should be migrated, no?

Comment (thienvh332) [reply queue_job/fields.py:27]: I will add it after fixing the pre-commit bug

**Comment by sbidoul** [comment]
In 18, warnings in logs cause tests to fail because they often reveal real issues. Two in this case:

```
errors that caused failure (2):
2024-10-07 11:47:52,343 245 WARNING odoo odoo.addons.queue_job.delay: Delayable Delayable(test.queue.job().testing_method((1,), {})) was prepared but never delayed 
2024-10-07 11:47:56,895 245 WARNING odoo odoo.addons.queue_job.utils: `queue_job__no_delay` ctx key found. NO JOB scheduled. 
```

**Comment by amh-mw** [comment]
Forward porting https://github.com/OCA/queue/pull/679 should handle `queue_job__no_delay`.

File: queue_job/tests/common.py [hunk 2]
```diff
@@ -444,9 +443,10 @@ def load_tests(loader, tests, ignore):
         Also extend the DocTestCase class trivially to fit the class teardown
         that Odoo backported for its own test classes from Python 3.8.
         """
-        if sys.version_info < (3, 8):
-            doctest.DocTestCase.doClassCleanups = lambda: None
-            doctest.DocTestCase.tearDown_exceptions = []
+        # UP036 Version block is outdated for minimum Python version
```
Comment (simahawk) [inline queue_job/tests/common.py]: do we need this?

**Review by simahawk — APPROVED** [review]
LGTM overall

**Comment by simahawk** [comment]
@thienvh332 reminder: if you don't quote w/ back-ticks the PR references, we get links automatically :wink:

**Review by sebalix — COMMENTED** [review]
Not sure we should port https://github.com/OCA/queue/pull/387 to revert it afterwards by porting https://github.com/OCA/queue/pull/466

**Comment by sebalix** [comment]
Should we also port these PRs?
- https://github.com/OCA/queue/pull/676
- https://github.com/OCA/queue/pull/683

They are missing when I check ports from 14.0 to 17.0

EDIT: fixed the 2nd PR ref

**Comment by simahawk** [comment]
> Not sure we should port #387 to revert it afterwards by porting #466

Good point. We should blacklist those PRs.

> Should we also port these PRs?

I'd say yes.

@thienvh332 ping

**Comment by thienvh332** [comment]
> > Not sure we should port #387 to revert it afterwards by porting #466

> Good point. We should blacklist those PRs.

In my opinion, these 2 commits have 1 more test case so we should keep it. I think the more test cases the better. Of course, let me know if you still want to ignore it. I will drop it.
> 
> > Should we also port these PRs?
> 
> I'd say yes.
> 
I have ported both of those commits. I missed adding them to the description.

**Comment by simahawk** [comment]
> In my opinion, these 2 commits have 1 more test case so we should keep it. I think the more test cases the better. Of course, let me know if you still want to ignore it. I will drop it.


ahh you are right, the PR was not fully reverted. Tests were kept. Fine for me :+1:

**Comment by sebalix** [comment]
To ease review, the migration commit is 21eb44d99a5a6bfbcce8ea1d8af2c508eba3f52f

Seems pre-commit stuff have been included in it, not sure?

**Comment by pedrobaeza** [comment]
@sebalix in this version, migrating from 17.0, there shouldn't be pre-commit changes, as most linters continue the same, the README files are already MD, and the pyproject.toml file remains the same.

**Comment by sbidoul** [comment]
> most linters continue the same

With the exception of ruff that was upgraded, and does some more python syntax upgrades plus some minor tweaks along blank lines multiple `with` context managers, etc.

**Review by sebalix — APPROVED** [review]

**Comment by simahawk** [comment]
@thienvh332 one last check

image

This change is included in the [migration commit](https://github.com/OCA/queue/pull/692/commits/21eb44d99a5a6bfbcce8ea1d8af2c508eba3f52f) but does not sounds like related to migration.
I guess is by error? Is it something we have to backport?

**Comment by thienvh332** [comment]
> This change is included in the [migration commit](https://github.com/OCA/queue/pull/692/commits/21eb44d99a5a6bfbcce8ea1d8af2c508eba3f52f) but does not sounds like related to migration. I guess is by error? Is it something we have to backport?

- This problem happened because the new test would raise an **`ERROR`** when a **`WARNING`** occurred in the test case. So I added it. 
- I separated it into a commit so I could backport it. (https://github.com/OCA/queue/pull/692/commits/d308edc4efe97df81ce61dfa840c9f5a587f57a4)

**Comment by sbidoul** [comment]
@thienvh332 please don't change the behaviour based on test_enabled. Better use `self.assertLogs` to trap the log and test that it actually logged what was expected. See for example: https://github.com/OCA/server-tools/pull/3050/commits/7d2e3d88c20175a91fcd5cf223417d9cf5da2f49

**Comment by thienvh332** [comment]
> @thienvh332 please don't change the behaviour based on test_enabled. Better use `self.assertLogs` to trap the log and test that it actually logged what was expected. See for example: [OCA/server-tools@7d2e3d8](https://github.com/OCA/server-tools/commit/7d2e3d88c20175a91fcd5cf223417d9cf5da2f49)

Hi @sbidoul 
I think you are right, I tried the update but it didn't work for me. Could you help me see what I am missing?

**Comment by thienvh332** [comment]
As [commented in unittest](https://github.com/OCA/queue/blob/c062b727ddfe984c68ef1bf4d16765e633da02d3/test_queue_job/tests/test_delayable.py#L282-L287), I am not able to control the printing of WARNING. So I will revert to the original handling. If you guys have any suggestions to handle it I will fix it.

**Comment by sbidoul** [comment]
The problem is we don't see what you tried anymore, so it's difficult to help.

**Comment by thienvh332** [comment]
Hi @sbidoul , indeed I should have pushed a fixup commit instead of force-pushing
- So here is what happened:
  - the **`.assertLogs`** was trapping the log as expected when calling **`node.__del__`** explicitely
  - but python was still garbage collecting the **`node`** after the test, so an additional warning was emitted
 
- To fix it, I now explicitly run the garbage collection with **`gc`**

Thanks for your help

**Comment by sbidoul** [comment]
@thienvh332 looks good, thanks!

File: test_queue_job/tests/test_delayable.py [hunk 3]
```diff
@@ -279,20 +282,23 @@ def test_delayable_group_of_chain(self):
         )
 
     def test_log_not_delayed(self):
-        logger_name = "odoo.addons.queue_job"
-        with self.assertLogs(logger_name, level="WARN") as test:
+        with self.assertLogs(level=logging.WARNING) as test:
```
Comment (mmequignon) [inline test_queue_job/tests/test_delayable.py:285]: Any WARNING log would be catched here, we should keep the logger filter.

File: test_queue_job/tests/test_job_auto_delay.py [hunk 4]
```diff
@@ -32,16 +32,7 @@ def test_auto_delay_inside_job(self):
 
     def test_auto_delay_force_sync(self):
         """method forced to run synchronously"""
-        with self.assertLogs(level="WARNING") as log_catcher:
-            result = (
-                self.env["test.queue.job"]
-                .with_context(_job_force_sync=True)
-                .delay_me(1, kwarg=2)
-            )
```
Comment (mmequignon) [inline test_queue_job/tests/test_job_auto_delay.py]: why aren't we keeping this `_job_force_sync` context key in the `test_auto_delay_force_sync` test method ?

**Review by mmequignon — COMMENTED** [review]
A few remarks, LG overall

File: test_queue_job/tests/test_job_auto_delay.py (hunk 4 above)
Comment (thienvh332) [reply test_queue_job/tests/test_job_auto_delay.py]: Hi @mmequignon ,
- Simone introduced that standardization in 16.0 (https://github.com/OCA/queue/pull/521). It should have been dropped as part of 17.0 migration but wasn't.
- So it's dropped in 18.0

**Comment by omalbastin** [comment]
Till v17 we can use the NothingtoDoJob exception to conditions that doesnot fall under a rule so that it will set the job to done. As the class NothingToDoJob has been removed from Exception in v18, how can we handle the same?

**Comment by john-herholz-dt** [comment]
@omalbastin did you find a way to replace it? We are encountering the same issue right now.

@sbidoul you might also give some hints here. Can you explain the removal of NothingToDoJob and whats the best way to replace it?

It is a bit unfortunate that most connector implementations are abandoned or in private forks - the framework itself is quite cool.

**Comment by sbidoul** [comment]
> Can you explain the removal of NothingToDoJob 

I'm afraid not. It seems it has been removed in this PR but I don't know why.

> and whats the best way to replace it?

The use case of that exception is unclear to me, to be honest.  If the job has nothing to do, why not simply... let it do nothing and end normally?

**Pull request merged**